{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metagenomics Project Code \n",
    "Author: Devon Shao  \n",
    "Date: 12/15/23\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#See Read.me for project description\n",
    "def map_reads_to_genomes(genomes, reads, k):\n",
    "    reads_dict = {}\n",
    "    num_of_genomes = len(genomes)\n",
    "    num_of_reads = len(reads)\n",
    "\n",
    "    # Iterate through reads, constructing a dict of dicts holding the first third and second third of the read sequence, the genome that the read matched with most closely and the smith_waterman score between the read and the section of the genome that the read matched most closely with\n",
    "    for i in range(num_of_reads):  \n",
    "        print(f'reads_dict construction: {i} / {num_of_reads}')\n",
    "        read_sequence = reads[i]  # Pull out the sequence of the read\n",
    "        first_third = read_sequence[0:k]\n",
    "        second_third = read_sequence[k:2*k]\n",
    "        reads_dict[read_sequence] = {'first_third':first_third, 'second_third':second_third, 'max_score':-999, 'genome_match':0}  \n",
    "\n",
    "    # Iterate through each genome, creating a hashmap for each genome where each 16-mer section of the genome is a key, and the position of its first nucleotide is the value. \n",
    "        # Then iterate through each read. Index the genome hashmap with the first 16 characters of each read to find the matching position in the genome, and use smith_waterman to calculate how well the overall read matches to that genome section. Do the samew with the second 16 characters of each read. \n",
    "    for j in range(num_of_genomes):\n",
    "        print(f\"genome_{j} / {num_of_genomes}\")\n",
    "        genome_sequence = genomes[j]\n",
    "        genome_hashmap = make_kmer_hashmap(genome_sequence, 16)\n",
    "        \n",
    "        # Iterate through each read \n",
    "        for k in range(num_of_reads):\n",
    "            read_sequence = reads[k]\n",
    "            first_third = reads_dict[read_sequence]['first_third']\n",
    "            second_third = reads_dict[read_sequence]['second_third']\n",
    "\n",
    "            if first_third in genome_hashmap:\n",
    "                positions = genome_hashmap[first_third]\n",
    "\n",
    "                for position in positions:\n",
    "                    smith_waterman_score = smith_waterman(read_sequence, genome_sequence[position:position + len(read_sequence)])\n",
    "\n",
    "                    if smith_waterman_score > reads_dict[read_sequence]['max_score']:\n",
    "                        reads_dict[read_sequence]['max_score'] = smith_waterman_score\n",
    "                        reads_dict[read_sequence]['genome_match'] = j\n",
    "            \n",
    "            elif second_third in genome_hashmap: \n",
    "                positions = genome_hashmap[second_third]\n",
    "\n",
    "                for position in positions:\n",
    "                    smith_waterman_score = smith_waterman(read_sequence, genome_sequence[position-k:position-k+len(read_sequence)])\n",
    "\n",
    "                    if smith_waterman_score > reads_dict[read_sequence]['max_score']:\n",
    "                        reads_dict[read_sequence]['max_score'] = smith_waterman_score\n",
    "                        reads_dict[read_sequence]['genome_match'] = j\n",
    "    return reads_dict\n",
    "\n",
    "\n",
    "# Function to construct a hashmap from a genome where the keys are all the k-mers of the genome and values are the positions in the genome where the k-mers begin.\n",
    "def make_kmer_hashmap(genome, k):\n",
    "    kmer_hashmap = {}\n",
    "    for kmer_start_pos in range(len(genome) - k + 1):\n",
    "        kmer_sequence = genome[kmer_start_pos:kmer_start_pos+k]\n",
    "        if kmer_sequence not in kmer_hashmap:\n",
    "            kmer_hashmap[kmer_sequence] = [kmer_start_pos]\n",
    "        else: \n",
    "            kmer_hashmap[kmer_sequence].append(kmer_start_pos)\n",
    "    return kmer_hashmap\n",
    "\n",
    "# This function takes two strings and returns a score that quantifies how good of a match those strings make.\n",
    "def smith_waterman(v:str, w:str):\n",
    "    s = [[0] * (len(w)+1) for _ in range(len(v)+1)]\n",
    "\n",
    "    for i in range(1, len(v)+1):\n",
    "        s[i][0] = 0\n",
    "    for j in range(1, len(w)+1):\n",
    "        s[0][j] = 0\n",
    "    for i in range(1,len(v)+1):\n",
    "        for j in range(1,len(w)+1):\n",
    "            match = 0\n",
    "            if v[i-1] == w[j-1]:\n",
    "                match = 1\n",
    "        \n",
    "            s[i][j] = max(s[i-1][j]-1, s[i][j-1]-1, s[i-1][j-1] + match)\n",
    "\n",
    "    return s[len(v)][len(w)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example use of the map_reads_to_genomes function, where you provide a list of genomes and a list of reads and k = 16\n",
    "\n",
    "#predictions = map_reads_to_genomes(genomes, reads, 16)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
